dkist/wcs/models.py

```diff
@@ -0,0 +1,340 @@
+    output of the right hand model's inverse output.
+
+    Parameters
+    ----------
+    op : `str`
+        The operator to use, can only be ``'&'``.
+    left : `astropy.modeling.Model`
+        The left hand model, should have one or more inputs which are shared
+        with the right hand model on the forward transform, and also rely on
+        these inputs for the inverse transform.
+    right : `astropy.modeling.Model`
+        The right hand model, no special behaviour is required here.
+    shared_inputs : `int`
+        The number of inputs (counted from the end of the end of the inputs to
+        the left model and the start of the inputs to the right model) which
+        are shared between the two models.
+
+    Example
+    -------
+
+    Take the following example with a time dependent celestial transform
+    (modelled as dependent upon the pixel coordinate for time rather than the
+    world coordinate).
+
+    The forward transform uses the "z" pixel dimension as input to both the
+    Celestial and Temporal models, this leads to the following transform in the
+    forward direction::
+
+        x  y           z
+        │  │           │
+        │  │  ┌────────┤
+        │  │  │        │
+        ▼  ▼  ▼        ▼
+      ┌─────────┐  ┌────────┐
+      │Celestial│  │Temporal│
+      └─┬───┬───┘  └───┬────┘
+        │   │          │
+        │   │          │
+        │   │          │
+        ▼   ▼          ▼
+       lon lat       time
+
+    This could trivially be reproduced using `~astropy.modeling.models.Mapping`.
+
+    The complexity is in the reverse transform, where the inverse Celestial
+    transform is also dependent upon the pixel coordinate z.
+    This means that the output of the inverse Temporal transform has to be
+    duplicated as an input to the Celestial transform's inverse.
+    This is achieved by the use of the ``Mapping`` models in
+    ``CoupledCompoundModel.inverse`` to create a multi-stage compound model
+    which duplicates the output of the right hand side model::
+
+       lon lat       time
+        │   │         │
+        │   │         ▼
+        │   │     ┌─────────┐
+        │   │     │Temporal'│
+        │   │     └──┬──┬───┘
+        │   │    z   │  │
+        │   │  ┌─────┘  │
+        │   │  │        │
+        ▼   ▼  ▼        │
+      ┌──────────┐      │
+      │Celestial'│      │
+      └─┬───┬────┘      │
+        │   │           │
+        ▼   ▼           ▼
+        x   y           z
+
+    """
+
+    _separable = False
+
+    def __init__(self, op, left, right, name=None, shared_inputs=1):
+        if op != "&":
+            raise ValueError(
+                f"The {self.__class__.__name__} class should only be used with the & operator."
+            )
+        super().__init__(op, left, right, name=name)
+        self.n_inputs = self.n_inputs - shared_inputs
+        self.inputs = self.inputs[:-shared_inputs]
```

Comment:
It's likely I don't understand how compound models work, but I was surprised to see the implicit assumption that the shared inputs are the last in the `self.inputs` list. Given the docstring above I expected them to be somewhere in the middle.

Reply:
The way the shared inputs work is that it's *all* the inputs to the right model also go into the left one, hence it's all the inputs at the end of the list.

Reply:
heh, which now I write out isn't actually enforced anywhere, I shall have a thing :smile:

Reply:
I have now changed it around so that it drops the leading inputs from the right model. Thanks for the spot!